Handle overlapping deployments gracefully

janetkuo opened the pull request:
Fixes #30028

<!-- Reviewable:start -->

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30730)

<!-- Reviewable:end -->

janetkuo committed:
Handle overlapping deployments gracefully

1. When overlapping deployments are discovered, annotate them
2. Expose those overlapping annotations as warnings in kubectl describe
3. Only respect the earliest updated one (skip syncing all other overlapping deployments)
4. Use indexer instead of store for deployment lister

AdoHe commented:
@janetkuo should we print the overlapping deployments names, thus user knows which deployment should be updated.

pwittrock commented:
@AdoHe want to review this with me?

janetkuo committed:
Add e2e test for overlapping deployments

janetkuo commented:
cc @kubernetes/deployment

AdoHe commented:
@pwittrock I am reviewing it now.

pwittrock commented:
Reviewed 1 of 4 files at r3.
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/controller/deployment/deployment_controller.go, line 434 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPo2RwH4390a4Iwcrt:-KPPo2RwH4390a4Iwcru:bailh1p) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/deployment_controller.go#L434)):*
> ```Go
> 	// overlaps, as well as all deployments that overlap this deployments, and sorting them.
> 
> 	// Relist all deployment for overlaps, and avoid syncing the old overlapping ones
> ```

I thought we avoid syncing the new  overlapping ones?

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPmyJ-Jj2A0zQUWiOF:-KPPmyJ09At6_yNejwm2:bueb6aa) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/util/deployment_util.go#L62)):*
> ```Go
> 	RollbackDone = "DeploymentRollback"
> 	// OverlapAnnotation marks deployments with overlapping selector with other deployments
> 	OverlapAnnotation = "deployment.kubernetes.io/overlap-with"
> ```

lets make this a little more in-your-face.  Maybe `deployment.kubernetes.io/error-overlapping-selectors` or `deployment.kubernetes.io/selector-overlap-error`

---

*[test/e2e/deployment.go, line 1196 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPQDKbEKNLmp0bhaC75:-KPQDKbEKNLmp0bhaC76:bxapcpj) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/test/e2e/deployment.go#L1196)):*
> ```Go
> 	d := newDeployment(deploymentName, replicas, podLabels, redisImageName, redisImage, extensions.RollingUpdateDeploymentStrategyType, nil)
> 	deploy, err := c.Extensions().Deployments(ns).Create(d)
> 	Expect(err).NotTo(HaveOccurred())
> ```

It is helpful to add a string here stating what you were trying to do when the error occurred.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

AdoHe commented:
*[pkg/controller/deployment/deployment_controller.go, line 456 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPQYWrek8WkyhemtHoh:-KPQYWrek8WkyhemtHoi:b-73m6vj) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L456)):*
> ```Go
> 		}
> 	}
> 	if !overlapping {
> ```

Nice catch.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

AdoHe commented:
*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPmyJ-Jj2A0zQUWiOF:-KPQYppOcJ7Dyjc0IdFl:bqoinr) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> lets make this a little more in-your-face.  Maybe `deployment.kubernetes.io/error-overlapping-selectors` or `deployment.kubernetes.io/selector-overlap-error`
</details>
Earlier I noticed this issue #30822 , maybe we should think more about this annotation.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -262,8 +264,12 @@ func (dc *DeploymentController) getDeploymentForReplicaSet(rs *extensions.Replic
+	if len(deployments) > 1 {
+		sort.Sort(byCreationTimestamp(deployments))
+		glog.Errorf("user error! more than one deployment is selecting replica set %s/%s with labels: %#v, returning %s/%s", rs.Namespace, rs.Name, rs.Labels, deployments[0].Namespace, deployments[0].Name)
+		glog.V(4).Infof("More than one deployment is selecting replica set %s/%s: %#v", rs.Namespace, rs.Name, deployments)
```
Is there any value to this log since you already log it above?

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -413,9 +430,56 @@ func (dc *DeploymentController) enqueueDeployment(deployment *extensions.Deploym
 	// overlaps, as well as all deployments that overlap this deployments, and sorting them.
+
+	// Relist all deployment for overlaps, and avoid syncing the newer overlapping ones (only sync the oldest one)
+	deployments, err := dc.dStore.List()
```
This is too expensive to happen on every enqueue. @deads2k got any ideas for fighting overlapping selectors here? How about indexing deployments with a StringSelectorIndex and skip enqueueing every time `dc.dStore.ByIndex(StringSelectorIndex, deployment.Spec.Selector)` returns deployments of the same namespace?

deads2k replied:
> This is too expensive to happen on every enqueue. @deads2k got any ideas for fighting overlapping selectors here? How about indexing deployments with a StringSelectorIndex and skip enqueueing every time dc.dStore.ByIndex(StringSelectorIndex, deployment.Spec.Selector) returns deployments of the same namespace?

I agree that this is too expensive, but only finding ones with the same index doesn't do the same thing since you can overlap without being the same.

Limiting to deployments in the same namespace and doing the check when work is going to be performed (on dequeue so that duplicate enqueues don't pay the tax multiple times) is probably enough.

deads2k commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -413,9 +430,56 @@ func (dc *DeploymentController) enqueueDeployment(deployment *extensions.Deploym
+	deployments, err := dc.dStore.List()
+	selector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		glog.Errorf("Deployment %s/%s has invalid label selector: %v", deployment.Namespace, deployment.Name, err)
```
`utilruntime.HandleError`

deads2k commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -413,9 +430,56 @@ func (dc *DeploymentController) enqueueDeployment(deployment *extensions.Deploym
+			// skip syncing this one if older overlapping one is found
+			shouldSkip := d.CreationTimestamp.Before(deployment.CreationTimestamp)
+			if shouldSkip {
+				glog.Errorf("Found deployment %s/%s has overlapping selector with deployment %s/%s, skip syncing it", deployment.Namespace, deployment.Name, d.Namespace, d.Name)
```
People who can see the controller logs aren't likely to mess with user deployments and users who have messed up their selectors aren't likely to be able to see the controller logs.  Seems like this should be either high verbosity or not at all.

deads2k commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -413,9 +430,56 @@ func (dc *DeploymentController) enqueueDeployment(deployment *extensions.Deploym
+			dc.markDeploymentOverlap(*deployment, d)
+			dc.markDeploymentOverlap(d, *deployment)
+			// skip syncing this one if older overlapping one is found
+			shouldSkip := d.CreationTimestamp.Before(deployment.CreationTimestamp)
```
Aren't I allowed update a selector? What if I update an older deployment to suddenly overlap?  Shouldn't the decision about what to do be based on which deployment has the oldest currently active RC?

AdoHe replied:
update selector should still be allowed. simply check creation timestamp is not enough to decide whether we can skip.

janetkuo commented:
Review status: 1 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/controller/deployment/deployment_controller.go, line 435 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-435:-KPTzR8V-EOpgrX-R-QB:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L435)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary>
> > This is too expensive to happen on every enqueue. @deads2k got any ideas for fighting overlapping selectors here? How about indexing deployments with a StringSelectorIndex and skip enqueueing every time dc.dStore.ByIndex(StringSelectorIndex, deployment.Spec.Selector) returns deployments of the same namespace?
> 
> 
> 
> I agree that this is too expensive, but only finding ones with the same index doesn't do the same thing since you can overlap without being the same.
> 
> 
> 
> Limiting to deployments in the same namespace and doing the check when work is going to be performed (on dequeue so that duplicate enqueues don't pay the tax multiple times) is probably enough.
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 438 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-438:-KPTzRun-3stptrRwGSe:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L438)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary>
> `utilruntime.HandleError`
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 448 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-448:-KPTzTwliA94ipZErtIW:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L448)):*
<details><summary><i>Previously, AdoHe (TonyAdo) wrote…</i></summary>
> update selector should still be allowed. simply check creation timestamp is not enough to decide whether we can skip.
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 450 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-450:-KPTzT5d1SwODH9gzGHd:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L450)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary>
> People who can see the controller logs aren't likely to mess with user deployments and users who have messed up their selectors aren't likely to be able to see the controller logs.  Seems like this should be either high verbosity or not at all.
</details>
Done.

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPmyJ-Jj2A0zQUWiOF:-KPTrfc42TnUgBSwncRd:biajynz) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, AdoHe (TonyAdo) wrote…</i></summary>
> Earlier I noticed this issue #30822 , maybe we should think more about this annotation.
</details>
This annotation is an inexpensive way for users to see they messed up the deployment selector (they can't see controller logs) (we can consume this in `kubectl describe`). Do you think we should hold on adding this annotation until we have a convention?

---

*[test/e2e/deployment.go, line 1196 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPQDKbEKNLmp0bhaC75:-KPU-6CcbxlLhKQebThF:b-qzrz2k) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/test/e2e/deployment.go#L1196)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> It is helpful to add a string here stating what you were trying to do when the error occurred.
</details>
Is the above `By()`enough to see where the error's from?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
Thank you all for the comments; code updated, ptal @pwittrock @AdoHe @kargakis @deads2k

pwittrock commented:
Review status: 1 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/controller/deployment/deployment_controller.go, line 435 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-435:-KPTZBCA-P1hIDFk-a0p:b-lroxen) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L435)):*
<details><summary><i>Previously, deads2k (David Eads) wrote…</i></summary>
> > This is too expensive to happen on every enqueue. @deads2k got any ideas for fighting overlapping selectors here? How about indexing deployments with a StringSelectorIndex and skip enqueueing every time dc.dStore.ByIndex(StringSelectorIndex, deployment.Spec.Selector) returns deployments of the same namespace?
> 
> 
> 
> I agree that this is too expensive, but only finding ones with the same index doesn't do the same thing since you can overlap without being the same.
> 
> 
> 
> Limiting to deployments in the same namespace and doing the check when work is going to be performed (on dequeue so that duplicate enqueues don't pay the tax multiple times) is probably enough.
</details>
Is this just hitting the cache and taking cpu time or does it actually perform rpcs?

---

*[pkg/controller/deployment/deployment_controller.go, line 448 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-448:-KPTYSzWL43EvKA6rGTU:b6plcoy) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L448)):*
<details><summary><i>Previously, AdoHe (TonyAdo) wrote…</i></summary>
> update selector should still be allowed. simply check creation timestamp is not enough to decide whether we can skip.
</details>
I agree consistently skipping the same deployment in the face of selector updates is better if it doesn't introduce additional complexity.  This PR is meant to have the control loop fail more gracefully when overlapping selectors are defined.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
Review status: 1 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/controller/deployment/deployment_controller.go, line 271 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-271:-KPU-Q_5XKiIRuqmBsSF:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L271)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> Is there any value to this log since you already log it above?
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 435 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-435:-KPUjBIHwaoylPPPaoNs:b-skhtnu) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L435)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Is this just hitting the cache and taking cpu time or does it actually perform rpcs?
</details>
It's just hitting the cache. 

---

*[pkg/controller/deployment/deployment_controller.go, line 448 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-448:-KPUjDDh4PgrVBu-vFFm:b-u00cqn) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L448)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> I agree consistently skipping the same deployment in the face of selector updates is better if it doesn't introduce additional complexity.  This PR is meant to have the control loop fail more gracefully when overlapping selectors are defined.
</details>
Did you mean just using deployment creationtimestamp is enough?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
Review status: 1 of 5 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/client/cache/listers.go, line 273 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPUisqBmdoyOOkYXPmo:-KPUisqBmdoyOOkYXPmp:boli9fa) ([raw file](https://github.com/kubernetes/kubernetes/blob/23e513e28e435cc798bc6c63d62566bc28cb7381/pkg/client/cache/listers.go#L273)):*
> ```Go
> }
> 
> func (s StoreToDeploymentNamespacer) List() (deployments []extensions.Deployment, err error) {
> ```

What does this do?  Add a comment for these fns

---

*[pkg/controller/deployment/deployment_controller.go, line 448 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-448:-KPUoJ58EjIX3MiKTXMa:bjcpuq1) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L448)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Did you mean just using deployment creationtimestamp is enough?
</details>
Actually thinking about this more, what does the time stamp of the RS give us?  If an overlap-update to the selectors or pod labels is made, then the Deployment will not create a new RS right?  Ideally the Deployment that was just updated to overlap would be the one that stopped working.  This would more closely match the expected behavior of the user - `I updated this and now it doesn't work` vs `I updated this and now a totally different object doesn't work`.

---

*[pkg/controller/deployment/deployment_controller.go, line 330 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPUj7naasQQqt_YOw8B:-KPUj7naasQQqt_YOw8C:bd3ugjq) ([raw file](https://github.com/kubernetes/kubernetes/blob/23e513e28e435cc798bc6c63d62566bc28cb7381/pkg/controller/deployment/deployment_controller.go#L330)):*
> ```Go
> 
> // getDeploymentForPod returns the deployment managing the ReplicaSet that manages the given Pod.
> // TODO: Surface that we are ignoring multiple deployments for a given Pod.
> ```

Update this comment with the new logic

---

*[pkg/controller/deployment/deployment_controller.go, line 340 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPUjMdAkkovHU5hWsx7:-KPUjMdAkkovHU5hWsx8:b-v9ribm) ([raw file](https://github.com/kubernetes/kubernetes/blob/23e513e28e435cc798bc6c63d62566bc28cb7381/pkg/controller/deployment/deployment_controller.go#L340)):*
> ```Go
> 		if len(rss) > 1 {
> 			sort.Sort(rsByCreationTimestamp(rss))
> 			glog.Errorf("user error! more than one replica set is selecting pod %s/%s with labels: %#v, only choose %s/%s", pod.Namespace, pod.Name, pod.Labels, rss[0].Namespace, rss[0].Name)
> ```

what does the 'only choose' mean?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
Review status: 1 of 5 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/controller/deployment/deployment_controller.go, line 448 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-448:-KPV3u1W-8hSJYdFChfL:b2m68nu) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L448)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Actually thinking about this more, what does the time stamp of the RS give us?  If an overlap-update to the selectors or pod labels is made, then the Deployment will not create a new RS right?  Ideally the Deployment that was just updated to overlap would be the one that stopped working.  This would more closely match the expected behavior of the user - `I updated this and now it doesn't work` vs `I updated this and now a totally different object doesn't work`.
</details>
Yeah thinking about this more, using active RS creationtimestamp (i.e. which deployment is updated last BEFORE I made this selector change) makes little difference with using deployment creationtimestamp -- it's pretty random. Shall we change it back to using creationtimestamp and we should eventually fix this with other ways, such as controller reference.

---

*[pkg/controller/deployment/deployment_controller.go, line 330 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPUj7naasQQqt_YOw8B:-KPUxx5XPDcpEN4gFB9J:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/23e513e28e435cc798bc6c63d62566bc28cb7381/pkg/controller/deployment/deployment_controller.go#L330)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> Update this comment with the new logic
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 340 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPUjMdAkkovHU5hWsx7:-KPV3jRp304yObEU3E0i:byis6rn) ([raw file](https://github.com/kubernetes/kubernetes/blob/23e513e28e435cc798bc6c63d62566bc28cb7381/pkg/controller/deployment/deployment_controller.go#L340)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> what does the 'only choose' mean?
</details>
I changed it to first sort rss by timestamp, and then loop over them until the first deployment is found

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

AdoHe commented:
*[pkg/client/cache/listers.go, line 273 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPUisqBmdoyOOkYXPmo:-KPV4w6fzBPhFrZukrFR:b-n2ueur) ([raw file](https://github.com/kubernetes/kubernetes/blob/23e513e28e435cc798bc6c63d62566bc28cb7381/pkg/client/cache/listers.go#L273)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> What does this do?  Add a comment for these fns
</details>
any need to return err here? It's always nil.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
Review status: 1 of 5 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/client/cache/listers.go, line 273 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPUisqBmdoyOOkYXPmo:-KPV5MdBbbjGS34XvUZG:bxzdgzp) ([raw file](https://github.com/kubernetes/kubernetes/blob/23e513e28e435cc798bc6c63d62566bc28cb7381/pkg/client/cache/listers.go#L273)):*
<details><summary><i>Previously, AdoHe (TonyAdo) wrote…</i></summary>
> any need to return err here? It's always nil.
</details>
Hm for now there's no error to return

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

AdoHe commented:
*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPmyJ-Jj2A0zQUWiOF:-KPV6k1lLzlawpzHuUBf:bnvj07c) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> This annotation is an inexpensive way for users to see they messed up the deployment selector (they can't see controller logs) (we can consume this in `kubectl describe`). Do you think we should hold on adding this annotation until we have a convention?
</details>
I think for now no need to hold on, we can change this in the future if we have a convention. I think `deployments.kubernetes.io/error-overlapping-selectors` could be better.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
Review status: 1 of 5 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPmyJ-Jj2A0zQUWiOF:-KPV86Eq68knDOX4O17S:b-dsz0h4) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, AdoHe (TonyAdo) wrote…</i></summary>
> I think for now no need to hold on, we can change this in the future if we have a convention. I think `deployments.kubernetes.io/error-overlapping-selectors` could be better.
</details>
If changed to this, the value would be the overlapping selectors, not the deployment's name?

But as a user, I already know what my deployment's selector is, and I'd be more interested in knowing which other deployment overlaps with mine. 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
Review status: 1 of 5 files reviewed at latest revision, 10 unresolved discussions.

---

*[test/e2e/deployment.go, line 1196 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPQDKbEKNLmp0bhaC75:-KPV8YH8FEi5litu7c15:be1kuek) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/test/e2e/deployment.go#L1196)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Is the above `By()`enough to see where the error's from?
</details>
The By helps, but isn't enough to distinguish between the different assertions beneath it.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
Review status: 1 of 5 files reviewed at latest revision, 10 unresolved discussions.

---

*[test/e2e/deployment.go, line 1196 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPQDKbEKNLmp0bhaC75:-KPVBIVOmITmpuFVEM7_:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/test/e2e/deployment.go#L1196)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> The By helps, but isn't enough to distinguish between the different assertions beneath it.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

AdoHe commented:
*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPmyJ-Jj2A0zQUWiOF:-KPVB7eB3a_itgv0Vsgl:b-dj96vi) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> If changed to this, the value would be the overlapping selectors, not the deployment's name?
> 
> But as a user, I already know what my deployment's selector is, and I'd be more interested in knowing which other deployment overlaps with mine. 
</details>
how about `error-selector-overlapping-with`? Also maybe we need to create a separate PR to consume this in kubectl describe?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
Reviewed 1 of 4 files at r5, 2 of 2 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPmyJ-Jj2A0zQUWiOF:-KPYhDl8nN7KPRJdnj-_:b-rf23bc) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, AdoHe (TonyAdo) wrote…</i></summary>
> how about `error-selector-overlapping-with`? Also maybe we need to create a separate PR to consume this in kubectl describe?
</details>
+1 to  making this highly visible in kubectl describe.  Feature freeze is Monday at midnight and it would be good to have it in by that date.

Lets go with:

`deployments.kubernetes.io/error-selector-overlapping-with`

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
Would you look into the test failure?

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-:-KPYiEdidiM1gsEMC-HL:bo7vnfo)*
<!-- Sent from Reviewable.io -->

kargakis commented on pkg/client/cache/listers.go:
```diff
@@ -265,6 +265,26 @@ func (s *StoreToReplicationControllerLister) GetPodControllers(pod *api.Pod) (co
 	return
 }
 
+type StoreToDeploymentNamespacer struct {
```
You should just use an Indexer instead of a Store in the `StoreToDeploymentLister` and then provide a `storeDeploymentsNamespacer` to pick the namespace you want. Look for example how the `storeReplicationControllersNamespacer` is implemented.

janetkuo commented:
The test failure is a known integration flake (not related to this change).

3 updates (late 3 commits):
1. Use indexer instead of store
2. Show warning message in kubectl describe on overlapping annotation 
3. Simply compare overlapping deployments by creation timestamp (removed comparing active RS timestamp)

---

Review status: 1 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/client/cache/listers.go, line 268 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r7-268:-KPZ-YJ5CZpM7Fblp0-r:bi7qlna) ([raw file](https://github.com/kubernetes/kubernetes/blob/65a36b6c0d1ceb9eefaa099db94ec6acef8afe66/pkg/client/cache/listers.go#L268)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> You should just use an Indexer instead of a Store in the `StoreToDeploymentLister` and then provide a `storeDeploymentsNamespacer` to pick the namespace you want. Look for example how the `storeReplicationControllersNamespacer` is implemented.
</details>
Thanks. Code updated to use indexerinformer also.

---

*[pkg/controller/deployment/deployment_controller.go, line 448 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r4-448:-KPZ-d_w-R9nlF6Re55L:bz4cwod) ([raw file](https://github.com/kubernetes/kubernetes/blob/32c25b9b116036289f742c165fd2e4ba4ef73f85/pkg/controller/deployment/deployment_controller.go#L448)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Yeah thinking about this more, using active RS creationtimestamp (i.e. which deployment is updated last BEFORE I made this selector change) makes little difference with using deployment creationtimestamp -- it's pretty random. Shall we change it back to using creationtimestamp and we should eventually fix this with other ways, such as controller reference.
</details>
Changed back to deployment creation timestamp 

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPmyJ-Jj2A0zQUWiOF:-KPZ2NCG_D8gWh8EKRKT:b-vvyu04) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> +1 to  making this highly visible in kubectl describe.  Feature freeze is Monday at midnight and it would be good to have it in by that date.
> 
> Lets go with:
> 
> `deployments.kubernetes.io/error-selector-overlapping-with`
</details>
Done. I was hesitating adding this before we finalize the annotation. Otherwise if we change it in the future, an older kubectl won't recognize the newer overlapping annotation. 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-:-KPZ-Sej5jl-gpx1fd9C:b-kmuzn5)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
@pwittrock @AdoHe @kargakis @deads2k  PTAL

kargakis commented on pkg/client/cache/listers.go:
```diff
@@ -261,14 +261,55 @@ func (s *StoreToReplicationControllerLister) GetPodControllers(pod *api.Pod) (co
+}
+
+// StoreToDeploymentNamespacer lists all deployments under its namespace in the store.
+func (s StoreToDeploymentNamespacer) List() (deployments []extensions.Deployment, err error) {
```
unexport this and add a label selector as an arg

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -262,8 +265,11 @@ func (dc *DeploymentController) getDeploymentForReplicaSet(rs *extensions.Replic
+	// trying to clean up one of the controllers, for now we just return the older one
+	if len(deployments) > 1 {
+		sort.Sort(byCreationTimestamp(deployments))
+		glog.Errorf("user error! more than one deployment is selecting replica set %s/%s with labels: %#v, returning %s/%s", rs.Namespace, rs.Name, rs.Labels, deployments[0].Namespace, deployments[0].Name)
```
remove the todo of this func

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -322,16 +328,25 @@ func (dc *DeploymentController) deleteReplicaSet(obj interface{}) {
-// TODO: Surface that we are ignoring multiple deployments for a given Pod.
+// If there are multiple deployments for a given Pod, only return the oldest one.
 func (dc *DeploymentController) getDeploymentForPod(pod *api.Pod) *extensions.Deployment {
 	rss, err := dc.rsStore.GetPodReplicaSets(pod)
```
Since it's about selectors why can't we go straight from a pod to its deployment? Feels like we should have something like dc.dStore.GetDeploymentsForPod.

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -491,6 +519,36 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 
+	// Handle overlapping deployments by deterministically avoid syncing deployments that fight over ReplicaSets.
+	// Relist all deployment in the same namespace for overlaps, and avoid syncing the newer overlapping ones (only sync the oldest one)
+	deployments, err := dc.dStore.Deployments(deployment.Namespace).List()
```
missing error check

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -491,6 +519,36 @@ func (dc *DeploymentController) syncDeployment(key string) error {
+			d, _ = dc.markDeploymentOverlap(d, other.Name)
+			other, _ = dc.markDeploymentOverlap(other, d.Name)
+			// Skip syncing this one if older overlapping one is found
+			// TODO: figure out a better way to determine which deployment to skip,
```
We could add an annotation on selector updates while the rest strategy is executed.

kargakis replied:
`deployment.kubernetes.io/selector-updated-at`:<timestamp>

kargakis commented on pkg/controller/deployment/util/deployment_util.go:
```diff
@@ -58,6 +58,8 @@ const (
 	// RollbackDone is the done rollback event reason
 	RollbackDone = "DeploymentRollback"
+	// OverlapAnnotation marks deployments with overlapping selector with other deployments
+	OverlapAnnotation = "deployment.kubernetes.io/error-selector-overlapping-with"
```
This feels more like a Deployment Condition for a permanently failed deployment ... apart from the fact that we still don't have perm-failed deployments:)

janetkuo commented:
Review status: 1 of 9 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/client/cache/listers.go, line 270 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r9-270:-KPZYYjRY_wtACZ-PNE-:b-67j1q9) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/client/cache/listers.go#L270)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> unexport this and add a label selector as an arg
</details>
Added selector, but can't unexport it (deployment controller uses it)

---

*[pkg/controller/deployment/deployment_controller.go, line 271 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r9-271:-KPZYXl2KEmztSiqqAJ2:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/deployment_controller.go#L271)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> remove the todo of this func
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 524 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r9-524:-KPZYnIIzXQBfLhxIDUK:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/deployment_controller.go#L524)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> missing error check
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 538 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r9-538:-KPZWMkKOh3uBdV1VmU0:b-u38d03) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/deployment_controller.go#L538)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> `deployment.kubernetes.io/selector-updated-at`: `<timestamp>`
</details>
When and where do we add this annotation?

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPdikASS8ED9quWv83-r9-62:-KPZZdA_UPXmxCZoBXqu:boxc0jy) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> This feels more like a Deployment Condition for a permanently failed deployment ... apart from the fact that we still don't have perm-failed deployments:)
</details>
Are you suggesting removing this annotation or renaming it?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

kargakis commented:
Review status: 1 of 9 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPdikASS8ED9quWv83-r9-62:-KPZaLogyKWoT5u6vyDQ:b8v16ch) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Are you suggesting removing this annotation or renaming it?
</details>
At this point, the annotation is fine. Once we have perm-failed deployments with conditions, we should consider transitioning this to a Condition.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

kargakis commented:
Review status: 1 of 9 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/controller/deployment/deployment_controller.go, line 538 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r9-538:-KPZazHAnVEGZ5XgWjW0:bu31jm2) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/deployment_controller.go#L538)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> When and where do we add this annotation?
</details>
pkg/registry/deployment/strategy.go#PrepareForUpdate every time the selector changes

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

kargakis commented:
Review status: 1 of 9 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/client/cache/listers.go, line 270 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r9-270:-KPZbhELaam8uyUvrIYO:b-3emdkb) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/client/cache/listers.go#L270)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Added selector, but can't unexport it (deployment controller uses it)
</details>
unexport the struct, not the method - similar to the rest of thr namespacers in this file

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
Review status: 1 of 9 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/client/cache/listers.go, line 270 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r9-270:-KP_0V9jF4jXwnTX8s0N:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/client/cache/listers.go#L270)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> unexport the struct, not the method - similar to the rest of thr namespacers in this file
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 333 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r9-333:-KP_0Xi0wWbsfPSMWMUT:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/deployment_controller.go#L333)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> Since it's about selectors why can't we go straight from a pod to its deployment? Feels like we should have something like dc.dStore.GetDeploymentsForPod.
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 538 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r9-538:-KP_0au7EfNiEvIYxXu-:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/deployment_controller.go#L538)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> pkg/registry/deployment/strategy.go#PrepareForUpdate every time the selector changes
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
Review status: 1 of 10 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPmyJ-Jj2A0zQUWiOF:-KP_FF04wCeIGhLOmwjC:b4upt6n) ([raw file](https://github.com/kubernetes/kubernetes/blob/9b1c55d545c9083c3edc0fd400b204021355330c/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Done. I was hesitating adding this before we finalize the annotation. Otherwise if we change it in the future, an older kubectl won't recognize the newer overlapping annotation. 
</details>
Good point.  Hopefully we won't have this problem too much farther into the future.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

AdoHe commented:
@janetkuo please fix build error:
```
pkg/controller/deployment/deployment_controller.go:584: k8s.io/kubernetes/pkg/api/unversioned.Time composite literal uses unkeyed fields
```

janetkuo commented:
Fixed the go-vet error; addressed all comments; added more test cases; ptal

kargakis commented on pkg/client/cache/listers.go:
```diff
@@ -261,14 +261,57 @@ func (s *StoreToReplicationControllerLister) GetPodControllers(pod *api.Pod) (co
 	return
 }
 
+type storeToDeploymentNamespacer struct {
```
nit: move this below `StoreToDeploymentLister`

kargakis commented on pkg/client/cache/listers.go:
```diff
@@ -316,6 +359,37 @@ func (s *StoreToDeploymentLister) GetDeploymentsForReplicaSet(rs *extensions.Rep
+		return
+	}
+
+	for _, m := range s.Indexer.List() {
```
Use `s.Deployments(pod.Namespace).List(labels.Everything())`

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -491,6 +507,39 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 		return dc.syncStatusOnly(d)
 	}
 
+	// Handle overlapping deployments by deterministically avoid syncing deployments that fight over ReplicaSets.
```
Can you extract all this diff into a `dc.isOverlapping` method in order to keep `syncDeployment` clean?

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -518,3 +567,45 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 	return fmt.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
+
+func selectorUpdatedBefore(d1, d2 *extensions.Deployment) bool {
```
All this diff should be part of the util package.

kargakis commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -518,3 +567,45 @@ func (dc *DeploymentController) syncDeployment(key string) error {
+func (o byCreationTimestamp) Len() int      { return len(o) }
+func (o byCreationTimestamp) Swap(i, j int) { o[i], o[j] = o[j], o[i] }
+func (o byCreationTimestamp) Less(i, j int) bool {
+	if o[i].CreationTimestamp.Equal(o[j].CreationTimestamp) {
```
You probably also want to take into account the `selector-updated-at` annotation too?

janetkuo commented:
Review status: 0 of 10 files reviewed at latest revision, 20 unresolved discussions.

---

*[pkg/client/cache/listers.go, line 264 \[r11\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r11-264:-KPjFm_T4W9CcUItrHON:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/98d10cd97004900b45a6331f545911e4dc4a690b/pkg/client/cache/listers.go#L264)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> nit: move this below `StoreToDeploymentLister`
</details>
Done.

---

*[pkg/client/cache/listers.go, line 371 \[r11\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r11-371:-KPjHcmYpUTrGQjUw0IH:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/98d10cd97004900b45a6331f545911e4dc4a690b/pkg/client/cache/listers.go#L371)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> Use `s.Deployments(pod.Namespace).List(labels.Everything())`
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 510 \[r11\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r11-510:-KPjP-SyWnR0_0AMhcs5:b-3rkvts) ([raw file](https://github.com/kubernetes/kubernetes/blob/98d10cd97004900b45a6331f545911e4dc4a690b/pkg/controller/deployment/deployment_controller.go#L510)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> Can you extract all this diff into a `dc.isOverlapping` method in order to keep `syncDeployment` clean?
</details>
Done (with `dc.handleOverlap`).

---

*[pkg/controller/deployment/deployment_controller.go, line 571 \[r11\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r11-571:-KPjPrnVJtK3vdMNPZKN:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/98d10cd97004900b45a6331f545911e4dc4a690b/pkg/controller/deployment/deployment_controller.go#L571)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> All this diff should be part of the util package.
</details>
Done.

---

*[pkg/controller/deployment/deployment_controller.go, line 596 \[r11\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r11-596:-KPjUQhoRbvPBUKuS-zC:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/98d10cd97004900b45a6331f545911e4dc4a690b/pkg/controller/deployment/deployment_controller.go#L596)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> You probably also want to take into account the `selector-updated-at` annotation too?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

kargakis commented on pkg/client/cache/listers.go:
```diff
@@ -316,6 +316,81 @@ func (s *StoreToDeploymentLister) GetDeploymentsForReplicaSet(rs *extensions.Rep
+	}
+	for _, d := range dList {
+		if len(pod.Labels[extensions.DefaultDeploymentUniqueLabelKey]) == 0 {
+			continue
```
break

kargakis replied:
Actually move this check outside of the loop before listing the deployments.

kargakis commented on pkg/client/cache/listers.go:
```diff
@@ -294,7 +294,7 @@ func (s *StoreToDeploymentLister) GetDeploymentsForReplicaSet(rs *extensions.Rep
 
 	// TODO: MODIFY THIS METHOD so that it checks for the podTemplateSpecHash label
-	for _, m := range s.Store.List() {
+	for _, m := range s.Indexer.List() {
```
`s.Deployments(pod.Namespace).List(labels.Everything())`

kargakis commented on pkg/client/cache/listers.go:
```diff
@@ -316,6 +316,81 @@ func (s *StoreToDeploymentLister) GetDeploymentsForReplicaSet(rs *extensions.Rep
+			return nil, fmt.Errorf("invalid label selector: %v", err)
+		}
+		// If a deployment with a nil or empty selector creeps in, it should match nothing, not everything.
+		if selector.Empty() || !selector.Matches(labels.Set(pod.Labels)) {
```
Can somebody else from @kubernetes/deployment also vet this? The relationship between replica sets and pods regarding selectors is obvious. What the relationship between 1) deployments and replica sets and subsequently 2) deployments and pods should really be? Up to this point, in order to determine that a pod belonged to a deployment, we listed all replica sets that manage the pod and then listed all deployments with selectors that match the **replica set labels**. So there is no guarantee that the deployment selector will match the pods matched by the matching replicas sets? Is the deployment selector destined to select amongst replica sets or should its selector be a part of hierarchy ie. it should be able to match all pods from all replica sets and replica set selectors should inherit and add on top the unique label?

pwittrock commented:
Review status: 0 of 10 files reviewed at latest revision, 23 unresolved discussions.

---

*[pkg/controller/deployment/deployment_controller.go, line 333 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r9-333:-KPnMfuX3sLC5CGPbNRF:b9h0c46) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/deployment_controller.go#L333)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Done.
</details>
In the future I would prefer to keep PRs as single purpose as possible and perform non-trivial restructuring / refactoring as a followup or precursor.  Doing it in the same PR will make it harder to debug bugs introduced as part of the restructuring or bug-fix.  

---

*[pkg/controller/deployment/deployment_controller.go, line 538 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r9-538:-KPn2xmw6VlJ5H2-PpMK:b-ovwaw7) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/deployment_controller.go#L538)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Done.
</details>
Good idea.

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPdikASS8ED9quWv83-r9-62:-KPn37gDMPznbhcCF4X3:b-aluu4c) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/util/deployment_util.go#L62)):*
nit: consider adding a comment explain precisely what this is used for / why it is set.  After we fix the overlapping selector issue we should clean up our code and delete these.

>
TODO: Delete this annotation when we gracefully handle overlapping selectors.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
*[pkg/controller/deployment/deployment_controller.go, line 333 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPLB7UVzhMXMXQwgqXs-r9-333:-KPnP0tQQZccLv3drEZ8:b-8nvwp6) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/deployment_controller.go#L333)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> In the future I would prefer to keep PRs as single purpose as possible and perform non-trivial restructuring / refactoring as a followup or precursor.  Doing it in the same PR will make it harder to debug bugs introduced as part of the restructuring or bug-fix.  
</details>
Very nice catch though.  Thanks @kargakis.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
Review status: 0 of 10 files reviewed at latest revision, 23 unresolved discussions.

---

*[pkg/client/cache/listers.go, line 297 \[r13\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r13-297:-KPnLsYfz88JyjwKF86t:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/20b1159a128f92dd336cbce1da8a3b9c6ccdc897/pkg/client/cache/listers.go#L297)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> `s.Deployments(rs.Namespace).List(labels.Everything())`
</details>
Done.

---

*[pkg/client/cache/listers.go, line 375 \[r13\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r13-375:-KPnLsDW-WiSLrwRzDBk:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/20b1159a128f92dd336cbce1da8a3b9c6ccdc897/pkg/client/cache/listers.go#L375)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> Actually move this check outside of the loop before listing the deployments.
</details>
Done.

---

*[pkg/client/cache/listers.go, line 383 \[r13\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r13-383:-KPnOvC76lunuGFwxLAN:bg839ve) ([raw file](https://github.com/kubernetes/kubernetes/blob/20b1159a128f92dd336cbce1da8a3b9c6ccdc897/pkg/client/cache/listers.go#L383)):*
<details><summary><i>Previously, kargakis (Michail Kargakis) wrote…</i></summary>
> Can somebody else from @kubernetes/deployment also vet this? The relationship between replica sets and pods regarding selectors is obvious. What the relationship between 1) deployments and replica sets and subsequently 2) deployments and pods should really be? Up to this point, in order to determine that a pod belonged to a deployment, we listed all replica sets that manage the pod and then listed all deployments with selectors that match the **replica set labels**. So there is no guarantee that the deployment selector will match the pods matched by the matching replicas sets? Is the deployment selector destined to select amongst replica sets or should its selector be a part of hierarchy ie. it should be able to match all pods from all replica sets and replica set selectors should inherit and add on top the unique label?
</details>
See [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/util/deployment_util.go#L385-L386): 

```go
func ListReplicaSets(deployment *extensions.Deployment, getRSList rsListFunc) ([]extensions.ReplicaSet, error) {
	// TODO: Right now we list replica sets by their labels. We should list them by selector, i.e. the replica set's selector
	//       should be a superset of the deployment's selector
```

Deployments should select their pods by their selectors. 

---

*[pkg/controller/deployment/util/deployment_util.go, line 62 \[r9\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPPdikASS8ED9quWv83-r9-62:-KPnQ6sd4efDPptn4KPY:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/277a5ce73e05beaf47bba476634aa8d9dc8a7ad4/pkg/controller/deployment/util/deployment_util.go#L62)):*
<details><summary><i>Previously, pwittrock (Phillip Wittrock) wrote…</i></summary>
> nit: consider adding a comment explain precisely what this is used for / why it is set.  After we fix the overlapping selector issue we should clean up our code and delete these.
> 
> >
> TODO: Delete this annotation when we gracefully handle overlapping selectors.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730)*
<!-- Sent from Reviewable.io -->

janetkuo commented:
ptal

kargakis commented on pkg/controller/deployment/util/deployment_util.go:
```diff
@@ -58,6 +58,12 @@ const (
 	// RollbackDone is the done rollback event reason
 	RollbackDone = "DeploymentRollback"
+	// OverlapAnnotation marks deployments with overlapping selector with other deployments
+	// TODO: Delete this annotation when we gracefully handle overlapping selectors.
```
How are we going to do this? Is there any issue open?

janetkuo replied:
With controller reference and/or auto-generated selectors, see #2210

kargakis commented on pkg/client/cache/listers.go:
```diff
@@ -278,40 +278,112 @@ func (s *StoreToDeploymentLister) Exists(deployment *extensions.Deployment) (boo
+			return nil, fmt.Errorf("invalid label selector: %v", err)
+		}
+		// If a deployment with a nil or empty selector creeps in, it should match nothing, not everything.
+		if selector.Empty() || !selector.Matches(labels.Set(rs.Labels)) {
```
I guess this is ok for now and more correct than previously since our utilities work in the same way but ultimately this can get out of hand especially with brave users. How about a follow-up issue to determine if we can maintain the deployment selector in its replica set labels even after users try to remove such a label?

janetkuo replied:
Sounds good

kargakis commented:
Two more questions, LGTM otherwise

pwittrock commented:
+lgtm 

---

Reviewed 1 of 7 files at r9, 4 of 5 files at r12, 1 of 3 files at r14, 4 of 4 files at r15.
Review status: all files reviewed at latest revision, 25 unresolved discussions.

---

*[pkg/client/cache/listers.go, line 305 \[r14\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-KPTzLbegKXTr7UBNSsw-r14-305:-KPsIPrOXBvDp3L-N_Yw:b-h54xi6) ([raw file](https://github.com/kubernetes/kubernetes/blob/1cf341251495decd328fd5e5b65e9090e921c586/pkg/client/cache/listers.go#L305)):*
<details><summary><i>Previously, janetkuo (Janet Kuo) wrote…</i></summary>
> Sounds good 
</details>
Good idea.  If this isn't in validation already we should add it later.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-:-KPtNUg890ONtcr-jbzX:bgc0spy)*
<!-- Sent from Reviewable.io -->

AdoHe commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30730#-:-KPtmy3nbRZANZQ78ge7:bnfp4nl)*
<!-- Sent from Reviewable.io -->

AdoHe commented:
Also please don't forget squash before merge.

janetkuo commented:
Yeah already squashed

janetkuo commented:
@k8s-bot node e2e test this issue: #31201
